packages/vendure-plugin-shipmate/src/api/shipmate.service.ts

```diff
@@ -0,0 +1,271 @@
+import { ShipmatePluginConfig } from '../shipmate.plugin';
+import { ShipmateConfigService } from './shipmate-config.service';
+import { parseOrder } from './util';
+import { generatePublicId } from '@vendure/core/dist/common/generate-public-id';
+import { FulfillOrderInput } from '@vendure/common/lib/generated-types';
+import { Response } from 'express';
+
+export const SHIPMATE_TOKEN_HEADER_KEY = 'X-SHIPMATE-TOKEN';
+export const SHIPMATE_API_KEY_HEADER_KEY = 'X-SHIPMATE-API-KEY';
+
+@Injectable()
+export class ShipmateService implements OnModuleInit {
+  tokens: Map<string, string> = new Map();
+  constructor(
+    private httpService: HttpService,
+    @Inject(PLUGIN_INIT_OPTIONS) private config: ShipmatePluginConfig,
+    private shipmateConfigService: ShipmateConfigService,
+    private orderService: OrderService,
+    private eventBus: EventBus,
+    private connection: TransactionalConnection,
+    private entityHydrator: EntityHydrator
+  ) {}
+
+  async onModuleInit(): Promise<void> {
+    this.httpService.axiosRef.defaults.headers.common['Accept'] =
+      'application/json';
+    this.httpService.axiosRef.defaults.headers.common['Content-Type'] =
+      'application/json';
+    this.eventBus.ofType(OrderPlacedEvent).subscribe(async ({ ctx, order }) => {
+      await this.setupRequestHeaders(ctx);
+      await this.createShipment(ctx, order);
+    });
+  }
+
+  async setupRequestHeaders(ctx: RequestContext) {
+    const shipmateConfig = await this.shipmateConfigService.getConfig(ctx);
+    if (!shipmateConfig) {
+      Logger.error(
+        `Shipmate credentials not configured for channel ${ctx.channel.code}`,
+        loggerCtx
+      );
+      return;
+    }
+    let shipmateTokenForTheCurrentChannel = this.tokens.get(
+      shipmateConfig.apiKey
+    );
+    if (!shipmateTokenForTheCurrentChannel) {
+      shipmateTokenForTheCurrentChannel = await this.getShipmentToken(
+        shipmateConfig.username,
+        shipmateConfig.password
+      );
+      this.tokens.set(shipmateConfig.apiKey, shipmateTokenForTheCurrentChannel);
+    }
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_TOKEN_HEADER_KEY
+    ] = shipmateTokenForTheCurrentChannel;
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_API_KEY_HEADER_KEY
+    ] = shipmateConfig.apiKey;
+  }
+
+  async getShipmentToken(
+    shipmateUsername: string,
+    shipmatePassword: string
+  ): Promise<string> {
+    const response = await firstValueFrom(
+      this.httpService.post<GetTokenRespose>(
+        `${this.config.shipmateApiUrl}/tokens`,
+        {
+          username: shipmateUsername,
+          password: shipmatePassword,
+        }
+      )
+    );
+    if (response.data.data?.token) {
+      Logger.info('Successfully authenticated with Shipmate API', loggerCtx);
+      return response.data.data.token;
+    } else {
+      Logger.error(response.data.message, loggerCtx);
+      throw new UserInputError(response.data.message);
```

Comment:
This one would go in the ShipMate client class

Reply:
I do see you are checking and throwing  response.data.message, this is Shipmate specific response data. This would be an argument to create a separate Client class though